docs/testing_mbed_OS_5.md

```diff
@@ -0,0 +1,187 @@
+
+```c++
+#include "mbed.h"
+#include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
+#include "rtos.h"
+
+using namespace utest::v1;
+
+// A test that returns successfully is considered successful
+void test_success() {
+    TEST_ASSERT(true);
+}
+
+// Tests that assert are considered failing
+void test_failure() {
+    TEST_ASSERT(false);
+}
+
+utest::v1::status_t test_setup(const size_t number_of_cases) {
+    // Setup Greentea using a reasonable timeout in seconds
+    GREENTEA_SETUP(40, "default_auto");
+    return verbose_test_setup_handler(number_of_cases);
+}
+
+// Test cases
+Case cases[] = {
+    Case("Testing success test", test_success),
+    Case("Testing failure test", test_failure),
+};
+
+Specification specification(test_setup, cases);
+
+// Entry point into the tests
+int main() {
+    return !Harness::run(specification);
+}
+```
+
+This test will first run a case that succeeds, then it will run a test that fails. This is a good template to use when creating tests. For more complex testing examples, please see the documentation for [utest](https://github.com/ARMmbed/mbed-os/tree/master/features/frameworks/utest).
+
+## Debugging tests
+
+Debugging tests is a crucial part of the development and porting process. This section will cover exporting the test, then driving the test with the test tools while the target is attached to a debugger.
+
+### Exporting tests
+
+Currently, the easiest way to export a test is to copy the test's source code from its test folder to your project's root. This way it will be treated like a normal application by the tools.
+
+You can find the path to the test you wish to export by running the following command:
+
+```
+mbed test --compile-list -n <test name>
+```
+
+Once you've copied all of the test's source files to your project root, go ahead and export your project:
+
+```
+mbed export -i <ide name>
+```
+
+Your exported project should now be in `projectfiles/<ide>_<target>`. Go ahead and open this project in your IDE.
+
+### Running a test while debugging
+
+Assuming your test has exported correctly to your IDE, go ahead and build the project and load it onto your target via your debugger.
+
+Bring the target out of reset and run the program. Your target will now be waiting for a synchronizing character string to be sent from the test tools over the serial port. The `mbed test` command cannot be run as that will attempt to flash the device, which you've already done with your IDE.
+
+Instead, the underlying test tools can be used to drive the test. [htrun](https://github.com/ARMmbed/htrun) is the tool that needs to be used in this case. This is installed when you install the requirements for mbed OS. However, if you don't have it installed you can do this by running `pip install mbed-host-tests`.
```

Comment:
It's part of the requirements for the tools (comes in with `pip install -r requirements.txt`), so if they have used the tools before they should have this installed. Do you think this part is redundant?